File: languages/kotlin/src/main/java/com/github/kavos113/karin/KarinLoader.java
package com.github.kavos113.karin;

import java.io.File;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;

class KarinLoader {
    private static boolean loaded = false;
    private static final String LIB_NAME = "karin_languages_java";

    public static synchronized void load() {
        if (loaded) {
            return;
        }

        String os = System.getProperty("os.name").toLowerCase();
        String resourcePath;

        if (os.contains("win")) {
            resourcePath = "/native/windows/" + LIB_NAME + ".dll";
        } else if (os.contains("mac")) {
            resourcePath = "/native/macos/lib" + LIB_NAME + ".dylib";
        } else if (os.contains("linux")) {
            resourcePath = "/native/linux/lib" + LIB_NAME + ".so";
        } else {
            throw new UnsupportedOperationException("Unsupported operating system: " + os);
        }

        try (InputStream inputStream = KarinLoader.class.getResourceAsStream(resourcePath)) {
            if (inputStream == null) {
                throw new RuntimeException("Failed to load native library from resource: " + resourcePath);
            }

            File tmpFile = File.createTempFile(LIB_NAME + "-", new File(resourcePath).getName());
            tmpFile.deleteOnExit();

            Files.copy(inputStream, tmpFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
            System.load(tmpFile.getAbsolutePath());
        } catch (Exception e) {
            throw new RuntimeException("Failed to load native library: " + e.getMessage(), e);
        } finally {
            loaded = true;
        }
    }
}

File: languages/kotlin/src/main/java/com/github/kavos113/karin/KarinLoader.kt
package com.github.kavos113.karin

import java.io.File
import java.nio.file.Files
import java.nio.file.StandardCopyOption

object KarinLoader {
    const val LIB_NAME = "karin_languages_java"

    private var loaded = false

    @Synchronized
    fun load() {
        if (loaded) {
            return
        }

        val os = System.getProperty("os.name").lowercase()
        val (dir, file) = when {
            os.contains("win") -> "windows" to "$LIB_NAME.dll"
            os.contains("mac") -> "macos" to "lib$LIB_NAME.dylib"
            os.contains("linux") -> "linux" to "lib$LIB_NAME.so"
            else -> throw UnsupportedOperationException("Unsupported operating system: $os")
        }

        val resourcePath = "/$dir/$file"

        val inputStream = KarinLoader::class.java.getResourceAsStream(resourcePath)
            ?: throw RuntimeException("Failed to load native library from resource: $resourcePath")

        val tmpFile = File.createTempFile("$LIB_NAME-", file)
        tmpFile.deleteOnExit()

        inputStream.use { input ->
            Files.copy(input, tmpFile.toPath(), StandardCopyOption.REPLACE_EXISTING)
        }

        System.load(tmpFile.absolutePath)
        loaded = true
    }
}